### PR TITLE
Functional extensions

### PR DESCRIPTION
**Pull request by gmrukwa**

### CODE REVIEW

**Comment by gmrukwa** [comment]
[Functional extensions](https://trello.com/c/mvcpywTs/219-functional-extensions)

File: src/Spectre.libPlatform/Transform.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+template< class InputType1, class InputType2, class OutputType, class BinaryOperation >
+std::vector<OutputType> transform(gsl::span<const InputType1> first, gsl::span<const InputType2> second, BinaryOperation binaryOperation, OutputType* = nullptr)
+{
+    if (first.size() == second.size())
+    {
+        std::vector<OutputType> result;
+        result.reserve(first.size());
+        // @gmrukwa: code below is equialent to the following line but causes no warning
```
Comment (Daroslaw) [inline src/Spectre.libPlatform/Transform.h]: Typo at `equialent`.

File: src/Spectre.libPlatform/Math.h [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+template< class DataType >
+constexpr std::vector<DataType> minus(gsl::span<const DataType> first, DataType second)
+{
+    return Functional::transform(first, [second](DataType left) { return left - second; });
+}
+
+template< class DataType >
+constexpr std::vector<DataType> times(gsl::span<const DataType> first, DataType second)
```
Comment (Daroslaw) [inline src/Spectre.libPlatform/Math.h]: "Times" may do its job, but if one seeks for multiplication operator "multiply" or "multiplyBy" may be better in my opinion.

File: src/Spectre.libPlatform/Transform.h [hunk 3]
```diff
@@ -67,7 +67,7 @@ std::vector<OutputType> transform(gsl::span<const InputType1> first, gsl::span<c
     {
         std::vector<OutputType> result;
         result.reserve(first.size());
-        // @gmrukwa: code below is equialent to the following line but causes no warning
+        // @gmrukwa: code below is equvialent to the following line but causes no warning
```
Comment (Daroslaw) [inline src/Spectre.libPlatform/Transform.h]: You had one job xd

File: src/Spectre.sln [hunk 4]
```diff
@@ -90,116 +90,189 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
...
+EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "Spectre.libPlatform.Tests", "Spectre.libPlatform.Tests\Spectre.libPlatform.Tests.vcxproj", "{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}"
+	ProjectSection(ProjectDependencies) = postProject
+		{76542909-5FC8-405B-8E6D-D2E529162A7A} = {76542909-5FC8-405B-8E6D-D2E529162A7A}
+	EndProjectSection
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
```
Comment (Sand3r-) [inline src/Spectre.sln]: ?

File: src/Spectre.sln [hunk 5]
```diff
@@ -90,116 +90,189 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
...
 		Release|x64 = Release|x64
+		Release|x86 = Release|x86
```
Comment (Sand3r-) [inline src/Spectre.sln]: ^

File: src/Spectre.sln [hunk 6]
```diff
@@ -90,116 +90,189 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
...
+		{76542909-5FC8-405B-8E6D-D2E529162A7A}.Release|x64.Build.0 = Release|x64
+		{76542909-5FC8-405B-8E6D-D2E529162A7A}.Release|x86.ActiveCfg = Release|x64
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Debug|x64.ActiveCfg = Debug|x64
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Debug|x64.Build.0 = Debug|x64
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Debug|x86.ActiveCfg = Debug|Win32
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Debug|x86.Build.0 = Debug|Win32
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Release|x64.ActiveCfg = Release|x64
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Release|x64.Build.0 = Release|x64
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Release|x86.ActiveCfg = Release|Win32
+		{81F10DE8-B9EF-400F-B3A7-57F2AFF35A14}.Release|x86.Build.0 = Release|Win32
```
Comment (Sand3r-) [inline src/Spectre.sln]: Are these 2 above necessary?

File: src/Spectre.libPlatform/Find.h [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        [&collection](size_t index) { return collection[index] != static_cast<DataType>(0); });
+    return preservedIndexes;
+}
+
+/// <summary>
+/// Finds the nonzero elements in specified collection.
+/// </summary>
+/// <param name="collection">The collection.</param>
+/// <returns>Vector with indexes of nonzero elements.</returns>
+std::vector<size_t> find(std::vector<bool> collection);
```
Comment (gmrukwa) [inline src/Spectre.libPlatform/Find.h]: It is implemented in corresponding `.cpp` file: `Find.cpp`.

File: src/Spectre.libPlatform/Math.h [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+// @gmrukwa: vector vs vector
+
+/// <summary>
+/// Add two vectors elementwise.
+/// </summary>
+/// <param name="first">The first.</param>
+/// <param name="second">The second.</param>
+/// <returns>Vector, elements of which are sums of corresponding elements in source vectors.</returns>
+template< class DataType >
+constexpr std::vector<DataType> plus(gsl::span<const DataType> first, gsl::span<const DataType> second)
```
Comment (gmrukwa) [inline src/Spectre.libPlatform/Math.h]: Right now, they're by means of std::transform. It is *as good as loops are* [source](https://www.youtube.com/watch?v=27tFyKc6g3U). Therefore, until we have some overpowered-low-level-black-magic-with-ultra-special-buffs, I'd stick to this. Either way - it's a single place for a substitution ;)

File: src/Spectre.libPlatform/Math.h [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+/// <summary>
+/// Multiply two vectors elementwise.
+/// </summary>
+/// <param name="first">The first.</param>
+/// <param name="second">The second.</param>
+/// <returns>Vector, elements of which are multiplications of corresponding elements in source vectors.</returns>
+template< class DataType >
+constexpr std::vector<DataType> multiplyBy(gsl::span<const DataType> first, gsl::span<const DataType> second)
```
Comment (gmrukwa) [inline src/Spectre.libPlatform/Math.h]: Nope. [Dot product](https://en.wikipedia.org/wiki/Dot_product) returns a *single number*. This doesn't. It's neither [cross product](https://en.wikipedia.org/wiki/Cross_product). This is just multiplication element-by-element, as stated in description. (Default for numpy's arrays or MATLAB's `.*`)

File: src/Spectre.sln (hunk 4 above)
Comment (gmrukwa) [inline src/Spectre.sln]: Good point.

File: src/Spectre.libPlatform/Math.h (hunk 9 above)
Comment (Sand3r-) [inline src/Spectre.libPlatform/Math.h]: Yeah, my bad.

**Pull request merged**